### ISSUE TITLE
ToDo list

### ISSUE THREAD

**Issue by jarun** [jarun/nnn#58]
Rolled from #1.

Feel free to suggest new features to add to the list below. Some of the features may still be master-only.

- [ ] ssh and scp support
- [ ] support terminal locking on OS X (any vlock replacement?)
- [x] run any number of scripts
- [x] prefer VISUAL over EDITOR
- [x] add a keybind to run a custom script (see #79)
- [x] integer `get_coolsize()` (thanks @arsv)
- [x] batch rename, move, delete files with vidir integration
- [x] copy file paths when X is unavailable
- [x] support copying multiple file names to copier
- [x] add open with... shortcut
- [x] add option to open bookmark directly
- [x] optimize memory usage and algorithms for file name handling
- [x] list and extract archives
- [x] show current selection number along with total entries in dir
- [x] support `NNN_NOWAIT` to unblock nnn (DE specific)
- [x] file and directory creation
- [x] change to light mode if number of cols < 35
- [x] in-place file rename
- [x] watch out for current dir changes (if not in filter or du mode)
- [x] pin and jump to a directory
- [x] show vol capacity and vol free in help
- [x] option to use exiftool instead of mediainfo
- [x] option to show dirs in custom color (default: enabled in blue)
- [x] show dir content size and number of files in du mode
- [x] support bookmarks string
- [x] add option to support opening text files in EDITOR
- [x] use `xdg-open` on Linux and `open(1)` on OS X as desktop opener
- [x] support multiple apps in `nlay`
- [x] use `vlock` as default terminal screensaver
- [x] *navigate-as-you-type* mode
- [x] subtree search (using gnome-search-tool or catfish)
- [x] support custom nlay
- [x] *search-as-you-type* at filter prompt
- [x] RPM packages for RHEL/CentOS (thanks @fennm)
- [x] support cd ..... (with . as PWD, stops at /)
- [x] jump to initial directory (where nnn was started)
- [x] `nlay` - a shell script for actions like open desktop search utility, run screensaver
- [x] show file details, mediainfo output, help in `less`
- [x] show media information (needs mediainfo)
- [x] support readline at chdir prompt
- [x] use `Q` to quit to current dir (needs external shell integration)
- [x] use `^Q` to quit (even from filter mode)
- [x] UTF-8 support
- [x] add option to start in disk usage analyzer mode
- [x] open current dir in desktop file manager
- [x] disk usage analyzer more
- [x] show unprivileged free blocks in du mode
- [x] shortcut `-` to jump to last directory
- [x] resolve `~` (HOME) at chdir prompt
- [x] option to show complete stats of just the current entry (name, file type, human readable permission string, byte size... all `lstat()` details)
- [x] update noice man page
- [x] sort entries by size
- [x] support file path copier script (`NNN_COPIER`)
- [x] support detail view (default)
  - [x] file type
  - [x] modification time
  - [x] human-readable file size
  - [x] current item in reverse video
  - [x] full name of currently selected file in bottom "bar"
- [x] roll navigation with `Up` and `Down` at the first and last entries
- [x] remove restriction on visiting parent dir when opened with a relative dir path
- [x] case insensitive alphabetic content listing
- [x] generate optimized binary
- [ ] ~show total number of files in du mode~ (apparently it has to be counted in nftw callback, with adjustments like file type check, hardlink check, which means unreasonable overhead)
- [ ] ~threaded/multicore nftw(3)~ (overkill, we are already very fast)
- [ ] ~use scandir instead of readdir~ (we are using custom alphanum sort, see https://github.com/jarun/nnn/commit/691291245a88d1c2277d8900a13f41b00685a72c)
- [ ] ~modify packagecore.yml and probably Makefile to generate OpenSUSE Leap package~

Anything else which would add value (please discuss in this thread)

**Comment by jarun** [comment]
@vaygr could you work on the second item in the list?

**Comment by vaygr** [comment]
Sure, I just need more insight on that. Do we mean so by marking several file names somehow and then copying them at once, separated by a newline character?

**Comment by jarun** [comment]
Exactly. I am thinking a (same one) keybind as start and stop and using space in between to _copy_ the selections.

Coming to using newline as the separator, please ensure one can use it with `rm` (using quotes around the complete copied string maybe?).

Also we'll need another keybind that works like pressed `Shift`. It would be great if we can press `Shift+space` twice and select all the entries *in between*.

**Comment by jarun** [comment]
Please share if you are thinking of any other approach...

**Comment by jarun** [comment]
@vaygr any update?

**Comment by 4goodapp** [comment]
Hi,
Can you add a shortcut to disable/enable " show directories first" please?
Specially in sort by time.

And if possible:

- [ ] Fuzzy filter (like **_fzf_** does).

- [ ] Files colored by type like, media, text, audio etc...
- [ ] Folder columns à la **_Ranger_** would be awesome too.

**Comment by jarun** [comment]
@4goodapp there are several reasons I chose to show directories on top in any of the 7 possible colors, the foremost is a superfast sort as dirs are eliminated immediately.

Fuzzy filter: we support regex. you can achieve what you want with `a.*b` [ref](https://www.codeproject.com/Articles/9099/The-Minute-Regex-Tutorial)

For the other requests, please take a look at [nnn design considerations](https://github.com/jarun/nnn/wiki/nnn-design-considerations) for a better explanation of the design decisions. And please stay open about the fact that _nnn is not ranger_.

**Comment by 4goodapp** [comment]
Alright, I keep **_nnn_** anyways just in case.

I like the way the filter work compared to _**ranger**_ .

Thanks. :)

**Comment by jarun** [comment]
Awesome! Have fun! 👍

**Comment by refacto** [comment]
It'd be nice to have a way to show the owner and group of files in the detail view.

**Comment by jarun** [comment]
@refacto it's already there:

Access: 0664/-rw-rw-r-- Uid: (1001/vaio)  Gid: (1001/vaio)

**Comment by refacto** [comment]
Ah, okay. `D` it is then. Thanks :)

**Comment by tom-i** [comment]
Hi @jarun, when are you going to do SSH and SCP support? It's quite big feature I think, btw.what about NFS, CIFS and so on.
Would be awesome to have such native support in your manager too :)

**Comment by jarun** [comment]
> It's quite big feature I think

They are. That's why I have added them in the ToDo list. Personally I don't have time to take those up.

**Comment by jtgd** [comment]
When I press ^K it briefly displays "copy not set" at the bottom and nothing is copied to the clipboard. I can't seem to find any information on this bug. Please advise.

**Comment by jarun** [comment]
https://github.com/jarun/nnn#copy-file-paths-to-clipboard

**Comment by jtgd** [comment]
I can't get 'R' to work with multiscript. I have this:

    export NNN_SCRIPT=/Home/jay/Bin/nscript
    export NNN_MULTISCRIPT=1

    -rwxr-xr-x 1 jay jay 112 Jul  4 04:08 /Home/jay/Bin/nscript
    -rwxr-xr-x 1 jay jay 105 Jul  4 03:45 /Home/jay/Bin/nscripttocp
    -rwxr-xr-x 1 jay jay 105 Jul  4 03:53 /Home/jay/Bin/nscripttomv

but when I type 'R' (single keystroke) nothing visible happens but it runs "nscript". There is no prompting for a suffix. Not sure what I'm doing wrong.

Also, you should consider passing the full filename of the current selection as $1 to the script. That would make it super more useful. As it is there's no way to make the script operate on a given file/dir.

**Comment by jarun** [comment]
Are you on master? The multiscript feature is not released yet.
Did you export NNN_MULTISCRIPT?

Yes, we can pass the param to the script. I will take a look into it.

**Comment by jarun** [comment]
Filename is passed as argument 1 at 042d00839245a3acf3cd0ce934f916a63dac4247.

**Comment by jarun** [comment]
Rolled at #110.
